staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go

```diff
@@ -125,10 +126,19 @@ func withLogging(handler http.Handler, stackTracePred StacktracePred, shouldLogR
 		rl := newLoggedWithStartTime(req, w, startTime)
 		rl.StacktraceWhen(stackTracePred)
 		req = req.WithContext(context.WithValue(ctx, respLoggerContextKey, rl))
-		defer rl.Log()
+		logInRoutine := false
+		defer func() {
+			if !logInRoutine {
+				rl.Log()
+			}
+		}()
 
 		w = responsewriter.WrapForHTTP1Or2(rl)
 		handler.ServeHTTP(w, req)
+
+		if routine.AppendTask(ctx, &routine.Task{Func: rl.Log}) {
+			logInRoutine = true
```

Comment:
can you explain a bit what would happen when `logInRoutine = true`?

Reply:
We added a feature to run the watch serving in a separate goroutine to free stack memory allocated by Golang runtime (https://github.com/kubernetes/kubernetes/pull/120902). This code determines if it is the case and run the `rl.log()` at the end of the aforementioned goroutine.

I also added comment to explain that a bit. Thanks!